### ISSUE TITLE
bleach is deprecated, html5lib unmaintained

### ISSUE THREAD

**Issue by pllim** [astropy/astropy#14316]
As per this announcement (repeated below so one less click for you):

* https://github.com/mozilla/bleach/issues/698

#### Summary

As of now, Bleach is deprecated.

We will continue to support Bleach:

* security updates
* support for new Python versions
* fixes for egregious bugs

I figure that's one release a year or something like that.

#### Why?

Bleach sits on top of--and heavily relies on--html5lib which is no longer in active development. It is increasingly difficult to maintain Bleach in that context and I think it's nuts to build a security library on top of a library that's not in active development. There are some options (switch to something else, take over html5lib, etc), I don't particularly like any of them. I think instead, someone new should explore the options with a brand new library and a fresh start.

# How does this affect astropy?

From our own installation doc:

* `html5lib`: To read `astropy.table.Table` objects from HTML files using the `pandas` reader.
* `bleach`: Used to sanitize text when disabling HTML escaping in the `astropy.table.Table` HTML writer.

Affected code:

https://github.com/astropy/astropy/blob/8d38984cbad87acce673f1885b0b5314494d6c0a/astropy/utils/xml/writer.py#L185

https://github.com/astropy/astropy/blob/8d38984cbad87acce673f1885b0b5314494d6c0a/astropy/io/ascii/html.py#L468

https://github.com/astropy/astropy/blob/8d38984cbad87acce673f1885b0b5314494d6c0a/astropy/io/ascii/html.py#L305

https://github.com/astropy/astropy/blob/8d38984cbad87acce673f1885b0b5314494d6c0a/astropy/io/misc/pandas/connect.py#L73-L74

Given both are optional dependencies, I think we can remove them but it would be API change and a lost of some features.

Should we wait and see how this plays out upstream? But how long?

**Comment by neutrinoceros** [comment]
Just stumbled upon this. Since bleach is needed for security reasons, we can't just vendor it (as I don't think we have the expertise to properly maintain it ourselves).

> Should we wait and see how this plays out upstream? But how long?

I think that's reasonable: removing the functionality altogether is probably not necessary as long as bleach continues to receive security fixes, but ideally we would just migrate to a better supported alternative.

In `XMLWriter.xml_cleaning_method` at least, bleach is exposed by name, so even if we change the internals to use something else, we'll need to evolve the API, so maybe we could start raising a `PendingDeprecationWarning` to raise awareness about this (and maybe get some help from people who need this functionality) ?

**Comment by pllim** [comment]
Not sure how visible `PendingDeprecationWarning` really is. Would like to hear from `io.ascii` and `io.misc` active maintainers (@taldcroft @hamogu @dhomeier @matteobachetti )

**Comment by matteobachetti** [comment]
I wouldn't mind putting some kind of warning about the uncertain future of the API. We should certainly guarantee some continuous support for I/O from html

**Comment by taldcroft** [comment]
I spent some time looking through the deprecation issue. From what I can see, the best alternative nh3/ammonia is not yet ready for prime time. This includes a recent memory leak issue.

I'm a little hesitant to put any sort of deprecation warning into astropy if there is absolutely no action that users can take based on that warning. At this point, what function does the warning serve apart from generating noise?

**Comment by taldcroft** [comment]
> A `PendingDeprecationWarning` would only be visible to downstream developers, not end users. I'm thinking it could bring attention from people that might be able to help.

I see your point. I just think it is wishful thinking that any developers using astropy will decide to help make `nh3` work better. I'm one such developer who uses astropy heavily, and the only thing I would do is update tests to silence that warning. Am I being pessimistic?

**Comment by pllim** [comment]
Yes, `PendingDeprecationWarning` is pretty unintrusive but TBH I am not sure even downstream developers can see it. It is a struggle to get some of them to even test against astropy nightly wheel.

**Comment by pllim** [comment]
(Does not hurt though to try. But only if we really going to upgrade that to DeprecationWarning at some point. There is no point to introduce a PendingDeprecationWarning and then just remove it if things don't work out.)

**Comment by neutrinoceros** [comment]
>  There is no point to introduce a PendingDeprecationWarning and then just remove it if things don't work out

That's why I proposed to do it specifically in `XMLWriter.xml_cleaning_method`, where we expose bleach by name, so my understanding is that we *have* to eventually deprecate this API since it'll stop working sooner or later.

**Comment by pllim** [comment]
Re: https://github.com/astropy/astropy/issues/14316#issuecomment-2252854108

Maybe a draft PR? 😸

**Comment by EmilStenstrom** [comment]
I think [JustHTML](https://github.com/EmilStenstrom/justhtml/blob/main/docs/sanitization.md) could be a good alternative to switch to?

**Comment by taldcroft** [comment]
Thanks for the heads-up @EmilStenstrom. This looks good at a first glance, I guess I only worry about long-term stability since it appears to be a one-person project at this point. But perhaps it only needs to be better than bleach, which is probably a low bar.

**Comment by EmilStenstrom** [comment]
@taldcroft You are right in that it's only me. Hoping to expand the community as the project gets more users. Not sure if it's any relief, but I'm not building this as a one-off and expect to keep it going for a long time.

The other strong alternative is nh3, which I think is also only one person. html5lib is unmaintained, and bs4 wraps stdlib's html.parser  which is not a html5 parser. lxml is not html5 either.

**Comment by taldcroft** [comment]
@EmilStenstrom - thanks for the update. We decided at the last astropy dev meeting to migrate to justHTML. If you have any interest in making the astropy PR for this (being much more expert in HTML sanitizing than anyone here), that would be great. Let me know, we will need to deprecate the existing `bleach_clean` cleaning method and add a new option for `justHTML`.

Otherwise no worries, this is in the queue of issues that @neutrinoceros will work on.

**Comment by EmilStenstrom** [comment]
@taldcroft Sorry, too much other stuff on my hands. Please open issues in the justhtml if you find stuff that would make your life easier. Happy to talk.

**Comment by neutrinoceros** [comment]
update: bleach got a ultimate release https://github.com/mozilla/bleach/issues/698#issuecomment-4631783739

To save a click: "Hello all! I released Bleach 6.4.0 which is the last release of Bleach. There will be no future releases including for Python support or security fixes. This is the end."

**Comment by pllim** [comment]
Should we just bite the bullet and deprecate them in astropy v9.0?

**Comment by neutrinoceros** [comment]
Okay I think I have a plan for dropping bleach, but I'm seeing differences between bleach's and justhtml's outputs and since HTML sanitising is really not my expertise so I'd rather ask first:

```py
import bleach
import justhtml

data = '<script>x</script> <em>OK</em>'

def raw(data: str) -> None:
    return data

def sanitize_bleach(data: str) -> None:
    return bleach.clean(data)

def sanitize_justhtml(data: str) -> None:
    return justhtml.JustHTML(data).to_html(pretty=False)

for f in [raw, sanitize_bleach, sanitize_justhtml]:
    print(f"{f.__name__}(data) = {f(data)!r}")
```

Here's the result
```
raw(data) = '<script>x</script> <em>OK</em>'
sanitize_bleach(data) = '&lt;script&gt;x&lt;/script&gt; <em>OK</em>'
sanitize_justhtml(data) = '<html><head> </head><body><em>OK</em></body></html>'
```
Is this an acceptable result, and is it equivalent ?

**Comment by neutrinoceros** [comment]
> Should we just bite the bullet and deprecate them in astropy v9.0?

That is actually *part* of my plan. But we certainly don't want to deprecate sanitizing altogether. Instead I'm thinking we could expose the functionality in a different way where users would be responsible for providing their own sanitizing functions. We'd be testing the ones we know about, of course, and we could recommend and document a `justhtml`-based solution.

**Comment by neutrinoceros** [comment]
(I'll study html5lib too before I even touch the code)

**Comment by neutrinoceros** [comment]
html5lib is actually never used directly. It's just an implementation detail that pandas needs it to read a table file with `flavor='bs4'`, but astropy should probably just not be concerned with this.
I'll open a PR shortly to resolve this part of the problem, and the way to go looks indeed like a simple deprecation.

**Comment by hamogu** [comment]
> Is this an acceptable result, and is it equivalent ?

It's not equivalent, but in this particular example, that's OK because the table reader wouldn't care about the parts that are different. It does mean that there also might be differences in other examples where it's inside a table, e.g. if a js script is associated with an element in a table cell. It's unlikely that there is any difference for simple, valid tables, but if users have built code on top of astropy (e.g. to mask cells that astropy reads in as string "&lt;script&gt;x&lt;/script&gt") then they might get a different result, since those cells would now be different.

It doesn't affect "normal" tables that the reader was build for (in particular parsing tables in ApJ, A&A, MNRAS html pages), and I don't think that can be avoided if we transition away from bleach. So, I think it's an "acceptable result".

**Comment by EmilStenstrom** [comment]
Hi! Author of justhtml here. There's a dedicated bleach migration doc that you can read if you want: https://emilstenstrom.github.io/justhtml/bleach-migration.html

@neutrinoceros The specific example you see if because of two options. 

1) To get things the most similar to bleach, use` JustHTML(html, fragment=True)`. This parses the document like a html fragement, not like a full document. This removes the wrapping html, head, body.

2) The other difference is what you want to do with dangerous tags like script. The default in justhtml is to remove them, the default in bleach is to keep them but escape their contents so they display as text on the page. If you're happy with removing them instead, that's the default, if you want to mirror bleach you create a small SanitizationPolicy and set it to escape. It's alll the migration guide.

Let me know if there's anything else I can help with!

**Comment by neutrinoceros** [comment]
Thank you so much. based on the migration guide I was able to get a smaller difference. Here's what I got so far

```py
import bleach
import justhtml
from collections.abc import Collection, Mapping
from typing import Any
import re

data = "<script>x</script> <em>OK</em>"


def raw(data: str) -> None:
    return data


def sanitize_bleach(data: str) -> None:
    return bleach.clean(data)


URL_LIKE_ATTRS = {
    "href",
    "src",
    "srcset",
    "poster",
    "action",
    "formaction",
    "data",
    "cite",
    "background",
    "ping",
}


def build_url_policy(
    allowed_tags: Collection[str],
    allowed_attributes: Mapping[str, Collection[str]],
) -> "justhtml.UrlPolicy":
    rules = {}

    global_attrs = allowed_attributes.get("*", ())
    for tag in allowed_tags:
        for attr in global_attrs:
            if attr in URL_LIKE_ATTRS:
                rules[(tag, attr)] = justhtml.UrlRule(
                    allowed_schemes={"http", "https", "mailto", "tel"},
                    allow_relative=True,
                )

    for tag, attrs in allowed_attributes.items():
        if tag == "*":
            continue
        for attr in attrs:
            if attr in URL_LIKE_ATTRS:
                rules[(tag, attr)] = justhtml.UrlRule(
                    allowed_schemes={"http", "https", "mailto", "tel"},
                    allow_relative=True,
                )
    return justhtml.UrlPolicy(allow_rules=rules)


ALLOWED_TAGS = frozenset(
    {
        "a",
        "abbr",
        "acronym",
        "b",
        "blockquote",
        "code",
        "em",
        "i",
        "li",
        "ol",
        "strong",
        "ul",
    }
)


def justhtml_clean_wrapper(
    html: str,
    *,
    tags: Collection[str] = ALLOWED_TAGS,
    attributes: Mapping[str, Collection[str]] | None = None,
    strip: bool = False,
    strip_comments: bool = True,
    # immediately unsupported kwargs (best effort)
    css_sanitizer: Any | None = None,
    protocols=None,
) -> str:
    attrs = dict(attributes or {})
    disallowed_tag_handling = "unwrap" if strip else "escape"

    # Bleach defaults to strip_comments=True. If you use escape mode
    # (`strip=False`) and still want comments removed rather than displayed,
    # remove comments before parsing.
    if strip_comments and disallowed_tag_handling == "escape":
        html = re.sub(r"<!--.*?-->", "", html, flags=re.DOTALL)

    if css_sanitizer is not None:
        raise TypeError("this isn't supported.")
    if protocols is not None:
        raise TypeError("this isn't supported.")

    policy = justhtml.SanitizationPolicy(
        allowed_tags=tags,
        allowed_attributes=attrs,
        allowed_css_properties=(),
        url_policy=build_url_policy(tags, attrs),
        drop_comments=strip_comments,
        disallowed_tag_handling=disallowed_tag_handling,
    )
    return justhtml.JustHTML(html, fragment=True, policy=policy).to_html(pretty=False)


def sanitize_justhtml_1(data: str) -> None:
    return justhtml_clean_wrapper(data)


def sanitize_justhtml_2(data: str) -> None:
    return justhtml_clean_wrapper(data, tags=())


for f in [raw, sanitize_bleach, sanitize_justhtml_1, sanitize_justhtml_2]:
    print(f"{f.__name__}(data) = {f(data)!r}")

```

output:

```
raw(data) = '<script>x</script> <em>OK</em>'
sanitize_bleach(data) = '&lt;script&gt;x&lt;/script&gt; <em>OK</em>'
sanitize_justhtml_1(data) = ' <em>OK</em>'
sanitize_justhtml_2(data) = ' &lt;em&gt;OK&lt;/em&gt;'
```

These results are still somewhat surprising to me, which I guess I'm not understanding `allowed_tags` correctly (as well, propably, as `allowed_attributes` ?).
I would expect `<script>x</script>`, which is not an allowed tag, to be escaped in `justhtml` under `disallowed_tag_handling='escape'`, just as it is in `bleach`.
I would actually expect `sanitize_justhtml_1(data)` to reproduce's `sanitize_bleach` exactly, but surprisingly, `sanitize_justhtml_2` gets closer to it with an empty `tags` (`allowed_tags`) argument.

**Comment by EmilStenstrom** [comment]
@neutrinoceros Sorry for the confusing docs, I'll see if I can update then to include the below additions:

I think the missing piece here is `drop_content_tags`. Some tags are configured to not only be removed by unwrapping (maintaining their content), but actually removed by deleting all their contents too. So simply set drop_content_tags=(), to skip that step in the policy.

The key backing detail is that JustHTML’s policy defaults `drop_content_tags` to `{"script", "style"}` and the sanitizer checks that before applying the generic `disallowed_tag_handling` branch.

sanitize_justhtml_2(data) isn't really closer, it's just escaping everything, since no tags are allowed.

**Comment by neutrinoceros** [comment]
Excellent, thanks a lot !
Now I have all the pieces to build a strictly backward compatible layer. What remains to be done is designing whatever the replacement API will be. I want to allow users to define their own sanitization strategies. justhtml will both be recommended and get first class support for that, I just don't want to make the same mistake of integrating a specific optional dependency too tightly within astropy. I don't think I need help with this part (yet ?), and I'm not sure exactly when I'll have focused time to work on it, but that's where I'm standing now.

**Comment by neutrinoceros** [comment]
FTR I'm tentatively planning to dedicate a couple days to this over July.

**Comment by gaborbernat** [comment]
One option that covers both replacements at once: turbohtml parses with the WHATWG algorithm (the html5lib job) and sanitizes against an allowlist (the bleach job) from one typed, maintained C library.

```python
from turbohtml.clean import sanitize

sanitize(untrusted_html)  # allowlist scrub; sanitize_report() returns what it removed
```

It is a clean-break API, not a bleach drop-in, so it would be a port rather than a swap; the docs have a bleach migration guide. https://turbohtml.readthedocs.io/en/stable/migration/bleach.html (Disclosure: I maintain it.)

**Comment by EmilStenstrom** [comment]
@neutrinoceros Just updated the bleach guide, the code snippet there should be as close to a drop in replacement as possible: https://emilstenstrom.github.io/justhtml/bleach-migration.html
